### PR TITLE
Issue #10188: Migrate Pitests from CircleCI to Github actions

### PR DESCRIPTION
**Pull request by AkMo3**
Resolves: #10188 

Current Configuration Timings:
1. GitHub actions: 9 min 
2. CircleCI: 18 min 

All these timings are for individual build times considering no backlogs are there.
Goals achieved:
1. All tests (including Pitests) run parallel now.
2. pitests are reafactored to not use parallelization on containers and we still have same parallelization on job level. 
2. Pitest artifacts are preserved BUT as zips - https://github.com/checkstyle/checkstyle/actions/runs/1044325180
image

Current issues: 
1. If Pitest fails, artifacts are not created for that specific pitest. Eg: [Pitest-main](https://github.com/checkstyle/checkstyle/runs/3101542141?check_suite_focus=true)
2. pitest-main fails due to unknown reasons. To be fixed in #9351. Till the time, suppression is being added.

### CODE REVIEW

**Comment by AkMo3** [comment]
@romani I am thinking of shifting all the Pitests to GitHub actions and shift the `no-exception` tests,  `no closed issue references  CI`, `check-pr-description` and `link_check` to Circle CI as they are of a total of 8 tests and we can run 4 concurrent tests as a configuration of (4 + 4) [2 Jobs], so as to reduce pressure on GitHub actions that has a limit of 20 concurrent jobs.

**Comment by romani** [comment]
from perspective of maintenance it is very beneficial to keep whole pitest on one CI, as it is easy what parts are critical and what are not. 

>  limit of 20 concurrent jobs.

if limit is just to make all jobs after 20th to be queued, it is completely ok. we have about 30 execution commands for pitest, so rest of them will execute after quick jobs finish their work.

**Comment by AkMo3** [comment]
> Javadoc execution takes about 15min, all other finish in few mins, but circleCI does not switch to new job execution untill finished.

We will not face this issue in Github Actions as all builds are independent. All VMs will run simultaneously.

All Pitests have been shifted to GitHub actions with artifacts preserved for 3 days. The no-exception tests have been shifted to CircleCI to reduce build time.

@strkkk Sir, can you please help with failing pitest2. It is identical to #9351, except we are running on macos-latest with a java 8 build.

File: .github/workflows/pitest_two.yml [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+          cp -r target/pit-reports/ pitest-javadoc-report/
+      - name: Archive code coverage results
+        uses: actions/upload-artifact@v2
+        with:
+          name: pitest2-javadoc-coverage-report
+          path: /home/runner/work/checkstyle/checkstyle/pitest-javadoc-report/
+          retention-days: 3
+
+  Pitest-indentation:
+    if: always()
```
Comment (strkkk) [inline .github/workflows/pitest_two.yml:63]: there is no need for this, it is true by default, isn't it?

File: .github/workflows/pitest_two.yml [hunk 2]
```diff
@@ -0,0 +1,106 @@
+name: "Pitest2"
```
Comment (strkkk) [inline .github/workflows/pitest_two.yml:1]: Please make name more meaningful, smth like "Pitest - main, javadoc and indentation packages"
Same for other files

Comment (strkkk) [inline .github/workflows/pitest_two.yml:1]: please also change method names to be more meaningful. Also, it may be better to put them all in one file named "Pitest.yml"

**Comment by strkkk** [comment]
> help with failing pitest2

I see now it is passing. I can recommend to try other macos images

**Comment by strkkk** [comment]
@AkMo3 please note - if you will move CI jobs here and there, badge links in readme can be broken

**Comment by AkMo3** [comment]
> I see now it is passing. I can recommend to try other macos images

It was passing since I had included a suppression in pitest.sh

> @AkMo3 please note - if you will move CI jobs here and there, badge links in readme can be broken

Yes, I had taken note of that. But, none of them should be affected as `no-exception` badge is not present, as circleCI batch depends if all tests pass or not, independent of what tests are.

File: .github/workflows/pitest_two.yml (hunk 1 above)
Comment (AkMo3) [reply .github/workflows/pitest_two.yml]: Yup, just wanted to be sure that all runs. Removed it now.

File: .github/workflows/pitest_two.yml (hunk 2 above)
Comment (AkMo3) [reply .github/workflows/pitest_two.yml]: Done

Comment (AkMo3) [reply .github/workflows/pitest_two.yml]: No, more required.

**Comment by AkMo3** [comment]
Current Configuration Timings:
1. GitHub actions: 22 min 
2. CircleCI : 18 min 

All these timings are for individual build times considering no backlogs are there.
All test (including Pitests) run parallel now.

Current issues: If Pitest fails, artifacts are not created for that specific pitest. Eg: [Pitest-main](https://github.com/checkstyle/checkstyle/runs/3101542141?check_suite_focus=true)

**Comment by strkkk** [comment]
have you tried to use image that was working ok on pitest as image for github action?

**Comment by AkMo3** [comment]
> have you tried to use image that was working ok on pitest as image for github action?

I was not able to find any specific image from CircleCI config, nor from any runs. It would be great if you can tell me which image it uses.

**Comment by strkkk** [comment]
https://circleci.com/docs/2.0/vm-service/#1-supply-amis there is ubuntu 16.04

File: .github/workflows/Pitests.yml [hunk 3]
```diff
@@ -0,0 +1,639 @@
...
+        uses: actions/checkout@v2
+      - name: pitest-common-2
+        run: |
+          mkdir pitest-common-2-report
+          ./.ci/pitest.sh pitest-common-2
+      - name: Copy result
+        if: always()
+        run: cp -r target/pit-reports/ pitest-common-2-report/
+      - name: Archive code coverage results
+        if: always()
```
Comment (strkkk) [inline .github/workflows/Pitests.yml:633]: there are still these `always()`, please remove all of them

File: .circleci/config.yml [hunk 4]
```diff
@@ -59,220 +63,103 @@ jobs:
...
-            else
-              echo "build is skipped ..."
-            fi
-      - store_artifacts:
-          path: ./target/pit-reports
+            sudo apt install curl
+            curl -s get.sdkman.io | bash
+            source "$HOME/.sdkman/bin/sdkman-init.sh"
+            sdk install groovy
+            groovy -version
```
Comment (strkkk) [inline .circleci/config.yml:143]: I think it is easier to just take image that has groovy pre-installed

File: .ci/pitest.sh [hunk 5]
```diff
@@ -47,6 +46,15 @@ pitest-annotation|pitest-design \
   checkPitestReport "${ignoredItems[@]}"
   ;;
 
+# Until Issue #9351 is fixed
```
Comment (strkkk) [inline .ci/pitest.sh:49]: please change comment to make it possible to check it by job, like

`till #9351`

**Review by strkkk — CHANGES_REQUESTED** [review]

File: .github/workflows/Pitests.yml [hunk 6]
```diff
@@ -0,0 +1,639 @@
...
+          key: checkstyle-maven-cache-${{ hashFiles('**/pom.xml') }}
+      - name: Checkout
+        uses: actions/checkout@v2
+      - name: Pitest-filters
+        run: |
+          mkdir pitest-filters-report
+          ./.ci/pitest.sh pitest-filters
+      - name: Copy result
+        if: always()
+        run: cp -r target/pit-reports/ pitest-filters-report/
```
Comment (nmancus1) [inline .github/workflows/Pitests.yml:581]: Why not just upload the report from `/target/pit-reports/` instead of copying?

File: .github/workflows/Pitests.yml (hunk 3 above)
Comment (AkMo3) [reply .github/workflows/Pitests.yml:533]: These are `always()` are in steps, not like previously which were on jobs. It is here to make sure that even if test fail, we can upload reports if any. These are not `true` by default, and must be manually set.

File: .circleci/config.yml (hunk 4 above)
Comment (AkMo3) [reply .circleci/config.yml:143]: But, the pre-built images like `circleci/buildpack-deps:groovy` or others don't work. It is showing the same error of `groovy: command not found` . Tried to install groovy on a docker image of `circleci/openjdk:stretch`, by `apt-get install groovy` but throws `404 not found` error on two packages. Seems this is the only way out currently.

File: .ci/pitest.sh (hunk 5 above)
Comment (AkMo3) [reply .ci/pitest.sh]: Done

File: .github/workflows/Pitests.yml (hunk 6 above)
Comment (AkMo3) [reply .github/workflows/Pitests.yml]: So that content inside every folder doesn't be pit-report but of which pitest name. Removed now.

Comment (AkMo3) [reply .github/workflows/Pitests.yml]: Please help with https://github.com/checkstyle/checkstyle/pull/10379#discussion_r674531564 . Need a better solution.

File: .circleci/config.yml (hunk 4 above)
Comment (strkkk) [reply .circleci/config.yml:143]: there is no need to use circleci images.
You can take any good official groovy image https://hub.docker.com/_/groovy

Comment (AkMo3) [reply .circleci/config.yml]: Used my own version of [groovy:jdk11](https://hub.docker.com/layers/groovy/library/groovy/jdk11/images/sha256-ffec94ae0c0eb1b2a906d92d81d0d4002f0a35e2df99da7d23a76c3053c392f3?context=explore), diff being added git and maven in container.

Comment (AkMo3) [reply .circleci/config.yml]: @strkkk Please complete the review

Comment (strkkk) [reply .circleci/config.yml]: What is the difference between your image and official groovy jdk11 image?

If there is none, it is better to use official one to avoid issues (e.g. if you delete image for some reason), if there are some difference (some files added, some setup, etc.) - please share dockerfile - add it to contribution repo with good comments

Comment (AkMo3) [reply .circleci/config.yml]: > What is the difference between your image and official groovy jdk11 image?

Only difference is that we have include maven and git in the image. We are using the official image from https://archive.apache.org/dist/groovy/

Please see https://github.com/checkstyle/contribution/pull/562/

File: .github/workflows/Pitests.yml [hunk 7]
```diff
@@ -0,0 +1,539 @@
...
+      - name: Archive code coverage results
+        if: always()
+        uses: actions/upload-artifact@v2
+        with:
+          name: pitest6-modifier-coverage-report
+          path: /home/runner/work/checkstyle/checkstyle/target/pit-reports/
+          retention-days: 3
+
+  Pitest-regexp:
+    if: always()
```
Comment (strkkk) [inline .github/workflows/Pitests.yml:329]: leftover

File: .github/workflows/Pitests.yml [hunk 8]
```diff
@@ -0,0 +1,539 @@
...
+      - name: Archive code coverage results
+        if: always()
+        uses: actions/upload-artifact@v2
+        with:
+          name: pitest6-blocks-coverage-report
+          path: /home/runner/work/checkstyle/checkstyle/target/pit-reports/
+          retention-days: 3
+
+  Pitest-sizes:
+    if: always()
```
Comment (strkkk) [inline .github/workflows/Pitests.yml:285]: same here

File: .github/workflows/Pitests.yml [hunk 9]
```diff
@@ -0,0 +1,539 @@
...
+        if: always()
+        uses: actions/upload-artifact@v2
+        with:
+          name: pitest6-sizes-coverage-report
+          path: /home/runner/work/checkstyle/checkstyle/target/pit-reports/
+          retention-days: 3
+
+  Pitest-modifier:
+    if: always()
+    runs-on: ubuntu-latest
```
Comment (strkkk) [inline .github/workflows/Pitests.yml:308]: and here

**Review by strkkk — COMMENTED** [review]
few minors

File: .circleci/config.yml (hunk 4 above)
Comment (strkkk) [reply .circleci/config.yml]: ok, that is fine.
the only thing that bugs me is that best place for this image is checkstyle docker repo.
@romani what do you think?

File: .github/workflows/Pitests.yml (hunk 7 above)
Comment (AkMo3) [reply .github/workflows/Pitests.yml]: Done

File: .github/workflows/Pitests.yml (hunk 8 above)
Comment (AkMo3) [reply .github/workflows/Pitests.yml]: Done

File: .github/workflows/Pitests.yml (hunk 9 above)
Comment (AkMo3) [reply .github/workflows/Pitests.yml:306]: Done

File: .circleci/config.yml (hunk 4 above)
Comment (AkMo3) [reply .circleci/config.yml]: > he only thing that bugs me is that best place for this image is checkstyle docker repo.

True, please use the dockerfile in contribution PR https://github.com/checkstyle/contribution/pull/562 .

**Review by strkkk — APPROVED** [review]

**Comment by strkkk** [comment]
@romani please take a look at discussion about docker image and contribution pr for it https://github.com/checkstyle/checkstyle/pull/10379#discussion_r676934468 
I think it is better to push to checkstyle image repo

**Comment by AkMo3** [comment]
Github, rebase

File: .circleci/config.yml [hunk 10]
```diff
@@ -1,278 +1,62 @@
 version: 2
 jobs:
 
-  build-caches:
-    machine: true
+  lucene-and-others-javadoc:
+    docker:
+      - image: akmo03/groovy-jdk11:latest
+    parallelism: 1
```
Comment (romani) [inline .circleci/config.yml:7]: Why we stop using ability to run separate commands in parallel ?

it should stay 4, and please use same hack by commands.txt to keep each job on 4 paralel executions.
We have limit of parallel jobs, but not a containers.
image

**Review by romani — CHANGES_REQUESTED** [review]
https://github.com/checkstyle/checkstyle/actions/runs/1044325180

image
this is strange, if we fail a build by diff validation, report should be created no matter what and be availbale for download.
Whole idea of reports to be available on failure to see details, on success nobody need such reports.
It is better to keep artifact on failure only. We can keep master build artifacts only in success (as prove that in master it is all green), but in PR, only for failed executions.


items:

File: .circleci/config.yml (hunk 10 above)
Comment (romani) [reply .circleci/config.yml:7]: hmmm https://app.circleci.com/pipelines/github/AkMo3/checkstyle/133/workflows/7b7b5a00-ef69-4e7d-8219-784743a3c9bc
looks like if we do not use continaers, we still use paralleization
image

Comment (romani) [reply .circleci/config.yml:7]: in this case usage of parallelzation in containers does not make benefit, lets keep config simple.

Comment (AkMo3) [reply .circleci/config.yml]: Done

**Comment by AkMo3** [comment]
> Do you know how to fix this ?

Fixed. Please see https://github.com/checkstyle/checkstyle/actions/runs/1090234042

File: .github/workflows/Pitests.yml [hunk 11]
```diff
@@ -0,0 +1,635 @@
...
+on:
+  push:
+    branches: [ master ]
+  pull_request:
+    branches: [ master ]
+
+jobs:
+
+  Pitest-main:
+    runs-on: macos-latest
```
Comment (romani) [inline .github/workflows/Pitests.yml:12]: why macos ? is here, please use ubuntu, it is our kind of default OS.
ALL other OSs are mostly checking "mvn verify" only.

File: .github/workflows/Pitests.yml [hunk 12]
```diff
@@ -0,0 +1,635 @@
...
+          key: checkstyle-maven-cache-${{ hashFiles('**/pom.xml') }}
+      - name: Checkout
+        uses: actions/checkout@v2
+      - name: Pitest-main
+        run: |
+          ./.ci/pitest.sh pitest-main
+      - name: Stage results
+        if: failure()
+        run: |
+          mkdir staging && cp -r target/pit-reports/ staging
```
Comment (romani) [inline .github/workflows/Pitests.yml:31]: please explain reason of copy report to special folder
if this sub folder in "target" is available in next step we can do archiving from it directly.

File: .circleci/config.yml [hunk 13]
```diff
@@ -1,278 +1,57 @@
...
-              echo "build is skipped ..."
-            fi
-      - store_artifacts:
-          path: ./target/pit-reports
+            ./.ci/validation.sh no-error-xwiki
 
 workflows:
   version: 2
-  pitest-testing:
+  No-Exception-Tests:
```
Comment (romani) [inline .circleci/config.yml:51]: lever case please

File: .ci/pitest.sh [hunk 14]
```diff
@@ -38,17 +38,25 @@ pitest-annotation|pitest-design \
 |pitest-packagenamesloader \
 |pitest-common-2|pitest-misc|pitest-xpath \
 |pitest-filters \
-|pitest-main \
 |pitest-coding \
 |pitest-regexp \
 |pitest-meta)
-  mvn -e -P$1 clean test org.pitest:pitest-maven:mutationCoverage;
+  mvn --no-transfer-progress -e -P$1 clean test org.pitest:pitest-maven:mutationCoverage;
```
Comment (romani) [inline .ci/pitest.sh:44]: please move to separate commit with prefix "minor: " (there are bunch of other commits like this)
we need to clearly separate changes for 101888 and other changes.

**Review by romani — CHANGES_REQUESTED** [review]
items:

File: .github/workflows/Pitests.yml [hunk 15]
```diff
@@ -0,0 +1,635 @@
+name: "Pitest"
```
[the remark this answers is not in the document]
Comment (AkMo3) [reply .github/workflows/Pitests.yml]: Done

File: .github/workflows/Pitests.yml [hunk 16]
```diff
@@ -0,0 +1,635 @@
...
+
+on:
+  push:
+    branches: [ master ]
+  pull_request:
+    branches: [ master ]
+
+jobs:
+
+  Pitest-main:
```
[the remark this answers is not in the document]
Comment (AkMo3) [reply .github/workflows/Pitests.yml]: Done

File: .github/workflows/Pitests.yml (hunk 11 above)
Comment (AkMo3) [reply .github/workflows/Pitests.yml]: Done.

File: .ci/pitest.sh (hunk 14 above)
Comment (AkMo3) [reply .ci/pitest.sh:44]: Done

File: .github/workflows/Pitests.yml (hunk 12 above)
Comment (AkMo3) [reply .github/workflows/Pitests.yml]: After a lot of testing, I found that we were not able to directly upload the pit-test folder if the tests fail. I did not found any concrete reason for it. GitHub docs suggest moving artifacts to a new folder and upload them. See: https://docs.github.com/en/actions/guides/building-and-testing-java-with-maven#packaging-workflow-data-as-artifacts

File: .circleci/config.yml [hunk 17]
```diff
@@ -1,278 +1,57 @@
 version: 2
 jobs:
 
-  build-caches:
-    machine: true
+  lucene-and-others-javadoc:
+    docker:
+      - image: akmo03/groovy-jdk8:latest
```
Comment (AkMo3) [inline .circleci/config.yml:6]: @romani Please let me know the details of the groovy image to checkstyle or your repo if pushed so that I can update hear

**Comment by romani** [comment]
@AkMo3 , please use checkstyle's image, I placed all details to https://github.com/checkstyle/contribution/pull/562

please resolve conflict and rebase to the latest code.

File: .github/workflows/pitests.yml [hunk 18]
```diff
@@ -0,0 +1,635 @@
...
+      - name: Stage results
+        if: failure()
+        run: |
+          mkdir staging && cp -r target/pit-reports/ staging
+      - name: Archive code coverage results
+        if: failure()
+        uses: actions/upload-artifact@v2
+        with:
+          name: pitest5-packagenamesloader-coverage-report
+          path: staging
```
Comment (romani) [inline .github/workflows/pitests.yml:137]: `retention-days: 3`

please add in all other places.

File: .github/workflows/pitests.yml [hunk 19]
```diff
@@ -0,0 +1,635 @@
...
+    runs-on: ubuntu-latest
+    steps:
+      - name: Setup local maven cache
+        uses: actions/cache@v2
+        with:
+          path: ~/.m2/repository
+          key: checkstyle-maven-cache-${{ hashFiles('**/pom.xml') }}
+      - name: Checkout
+        uses: actions/checkout@v2
+      - name: pitest-Common
```
Comment (romani) [inline .github/workflows/pitests.yml:246]: pitest-design

File: .github/workflows/pitests.yml [hunk 20]
```diff
@@ -0,0 +1,635 @@
...
```
Comment (romani) [inline .github/workflows/pitests.yml:446]: pitest-common

File: .github/workflows/pitests.yml [hunk 21]
```diff
@@ -0,0 +1,635 @@
...
+          mkdir staging && cp -r target/pit-reports/ staging
+      - name: Archive code coverage results
+        if: failure()
+        uses: actions/upload-artifact@v2
+        with:
+          name: pitest4-header-coverage-report
+          path: staging
+          retention-days: 3
+
+  pitest-common-2:
```
Comment (romani) [inline .github/workflows/pitests.yml:612]: please place it close to "pitest-common"

**Review by romani — CHANGES_REQUESTED** [review]
please reuse https://github.com/checkstyle/checkstyle/blob/master/.github/workflows/pitest.yml
and update it up to your latest "tempale" of just jobs.


items:

File: .circleci/config.yml (hunk 4 above)
Comment (romani) [reply .circleci/config.yml]: docker image uploaded to checkstyle dockerhub

File: .ci/jsoref-spellchecker/whitelist.words [hunk 22]
```diff
@@ -23,6 +23,7 @@ AFDF
 AFilter
 aggregators
 Agoogle
+akmo
```
Comment (romani) [inline .ci/jsoref-spellchecker/whitelist.words:26]: Why we need this ?

File: .github/workflows/pitest.yml [hunk 23]
```diff
@@ -7,6 +7,636 @@ on:
   pull_request:
 
 jobs:
+  pitest-main:
```
Comment (romani) [inline .github/workflows/pitest.yml:10]: `name: pitest`
Please rename of workflow, as you had

**Review by romani — CHANGES_REQUESTED** [review]
Items:

File: .github/workflows/pitests.yml [hunk 24]
```diff
@@ -0,0 +1,635 @@
...
+        if: failure()
+        run: |
+          mkdir staging && cp -r target/pit-reports/ staging
+      - name: Archive code coverage results
+        if: failure()
+        uses: actions/upload-artifact@v2
+        with:
+          name: pitest2-javadoc-coverage-report
+          path: staging
+          retention-days: 3
```
[the remark this answers is not in the document]
Comment (AkMo3) [reply .github/workflows/pitests.yml]: We cannot make a step contain both `use` and `run`. Eg: https://github.com/checkstyle/checkstyle/actions/runs/1120812524

File: .ci/jsoref-spellchecker/whitelist.words (hunk 22 above)
Comment (AkMo3) [reply .ci/jsoref-spellchecker/whitelist.words:26]: This was included by some rebasing error. Removed.

File: .github/workflows/pitest.yml (hunk 23 above)
Comment (AkMo3) [reply .github/workflows/pitest.yml:10]: May be "Pitest" as other starts with caps too.

File: .github/workflows/pitests.yml (hunk 21 above)
Comment (AkMo3) [reply .github/workflows/pitests.yml]: Done

File: .github/workflows/pitests.yml (hunk 20 above)
Comment (AkMo3) [reply .github/workflows/pitests.yml]: Done

File: .github/workflows/pitests.yml (hunk 19 above)
Comment (AkMo3) [reply .github/workflows/pitests.yml]: Done

File: .github/workflows/pitests.yml (hunk 18 above)
Comment (AkMo3) [reply .github/workflows/pitests.yml]: Done

File: .github/workflows/pitests.yml (hunk 24 above)
[the remark this answers is not in the document]
Comment (AkMo3) [reply .github/workflows/pitests.yml]: Done

File: .github/workflows/pitests.yml [hunk 25]
```diff
@@ -0,0 +1,635 @@
...
+          ./.ci/pitest.sh pitest-main
+      - name: Stage results
+        if: failure()
+        run: |
+          mkdir staging && cp -r target/pit-reports/ staging
+      - name: Archive code coverage results
+        if: failure()
+        uses: actions/upload-artifact@v2
+        with:
+          name: pitest2-main-coverage-report
```
[the remark this answers is not in the document]
Comment (AkMo3) [reply .github/workflows/pitests.yml]: Done

File: .circleci/config.yml [hunk 26]
```diff
@@ -1,278 +1,57 @@
...
-      # we do not do thorough testing of gui part
-      # - pitest7:
-      #     requires:
-      #       - build-caches
-      - pitest8:
-          requires:
-            - build-caches
+      - lucene-and-others-javadoc
+      - cassandra-storm-tapestry-javadoc
+      - hadoop-apache-groovy-scouter-javadoc
```
[the remark this answers is not in the document]
Comment (AkMo3) [reply .circleci/config.yml]: Done

File: .ci/jsoref-spellchecker/whitelist.words [hunk 27]
```diff
@@ -764,6 +765,7 @@ lucene
 lui
 luiframework
 LVL
+macos
```
[the remark this answers is not in the document]
Comment (AkMo3) [reply .ci/jsoref-spellchecker/whitelist.words]: Done

File: .github/workflows/pitest.yml (hunk 23 above)
Comment (romani) [reply .github/workflows/pitest.yml:10]: Ok, let s align with others.

**Review by romani — CHANGES_REQUESTED** [review]
Please alphabetically order jobs in yml file, to ease navigation letter on.

**Review by romani — APPROVED** [review]
Very good now
